### PR TITLE
Creating turf from map

### PR DESCRIPTION
**Pull request by firechicago**
Basic implementation of the turf cutter tool

### CODE REVIEW

File: app/controllers/lists_controller.rb [hunk 1]
```diff
@@ -15,6 +15,10 @@ def create
     end
   end
 
+  def index
+    @lists = List.all
```
Comment (dpickett) [inline app/controllers/lists_controller.rb]: this should be paginated or limited to ensure the collection of lists do not exceed memory limits.

File: app/controllers/turfs_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+class TurfsController < ApplicationController
+  def create
+    unless current_user
```
Comment (dpickett) [inline app/controllers/turfs_controller.rb]: place this in a `before_action` to create more clarity as to the intent of the `create` method

File: app/controllers/turfs_controller.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      flash[:notice] = "You must be signed in to do that."
+      render "/"
+      return
+    end
+    params["lists"].each_value do |list|
```
Comment (dpickett) [inline app/controllers/turfs_controller.rb]: what happens if the params["lists"] is left unspecified? Hint: it will be `nil`.

File: app/controllers/turfs_controller.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      @list = List.new
+      @list.name = list["name"]
+      if @list.save
+        list["voter_ids"].each do |voter_id|
```
Comment (dpickett) [inline app/controllers/turfs_controller.rb]: you're assuming a certain data structure in the params hash. You can't really assume the params hash will always be submitted in the way you intend it to. Your code should be defensive so it properly takes this into account.

Comment (dpickett) [inline app/controllers/turfs_controller.rb]: consider creating a plain old ruby object using unit tests to manage this operation. Taking it out of the controller makes your behaviors easier to test, and keeps your controller skinny.

File: app/controllers/turfs_controller.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+          ListMembership.create(voter_id: voter_id, list_id: @list.id)
```
Comment (dpickett) [inline app/controllers/turfs_controller.rb]: what happens if the `ListMembership` is invalid? Consider `create!` or managing the use case where a ListMembership would be invalid

Comment (dpickett) [inline app/controllers/turfs_controller.rb]: assign the associated objects directly. Consider

```ruby
@list.memberships.build do |m|
  m.voter = voter #you need to supply this object
end
```

Comment (dpickett) [inline app/controllers/turfs_controller.rb]: this assumes that the voter_id will exist in the table. You should use `list['voter_ids']` to find the voters in the table, and loop directly through those objects instead

File: app/views/lists/map_view.html.erb [hunk 6]
```diff
@@ -37,16 +37,42 @@ var voters = <%= @geojson.html_safe %>
...
+map.on('draw:created', findVoters);
 
-function showPolygonArea(e) {
+function findVoters(e) {
   var new_layer = map.addLayer(e.layer);
-  if (leafletPip.pointInLayer(marker.geometry.coordinates, new_layer).length > 0) {
-    alert("The point is in one of your shapes");
-  } else {
-    alert("The point is outside your shapes");
+  array = []
```
Comment (dpickett) [inline app/views/lists/map_view.html.erb]: semicolon

Comment (dpickett) [inline app/views/lists/map_view.html.erb]: avoid variable names like `array` - can you utilize a better name?

File: app/views/lists/map_view.html.erb [hunk 7]
```diff
@@ -37,16 +37,42 @@ var voters = <%= @geojson.html_safe %>
...
+var lists = [];
+
 map.fitBounds(votersLayer.getBounds());
 
-map.on('draw:created', showPolygonArea);
+map.on('draw:created', findVoters);
 
-function showPolygonArea(e) {
+function findVoters(e) {
   var new_layer = map.addLayer(e.layer);
```
Comment (dpickett) [inline app/views/lists/map_view.html.erb]: idiomatically, the variable name should be `newLayer`

Comment (dpickett) [inline app/views/lists/map_view.html.erb]: can you consider a better variable name for this layer?

File: app/views/lists/map_view.html.erb [hunk 8]
```diff
@@ -37,16 +37,42 @@ var voters = <%= @geojson.html_safe %>
...
   };
-}
+  lists.push(array)
+};
+
+$(document).ready(function(){
+  $(".save").click(function(e){
+    e.preventDefault();
+    var listsName = prompt("Enter a name for these lists");
+    var listsData = []
```
Comment (dpickett) [inline app/views/lists/map_view.html.erb]: semicolo

File: app/views/lists/map_view.html.erb [hunk 9]
```diff
@@ -37,16 +37,42 @@ var voters = <%= @geojson.html_safe %>
...
+      for (var j = 0; j < list.length; j++) {
+        voter = list[j];
+        voterIDs.push(voter.properties.voter_id);
+      };
+      listsData.push({ name: listName, voter_ids: voterIDs });
+    }
+    $.ajax({type: "POST",
+      url: "/map_lists_create",
+      data: { lists: listsData }
+    })
```
Comment (dpickett) [inline app/views/lists/map_view.html.erb]: you should have success and error callbacks on this AJAX routine to provide the user with feedback that the save successful or not successful.

File: app/views/lists/map_view.html.erb [hunk 10]
```diff
@@ -37,16 +37,42 @@ var voters = <%= @geojson.html_safe %>
...
+      var listName = listsName + " Turf " + (i + 1);
+      var voterIDs =[];
+      for (var j = 0; j < list.length; j++) {
+        voter = list[j];
+        voterIDs.push(voter.properties.voter_id);
+      };
+      listsData.push({ name: listName, voter_ids: voterIDs });
+    }
+    $.ajax({type: "POST",
+      url: "/map_lists_create",
```
Comment (dpickett) [inline app/views/lists/map_view.html.erb]: consider a more restful resource. You are creating map_lists so why not post to `/map_lists`?

File: coverage/.resultset.json [hunk 11]
```diff
@@ -37,6 +37,7 @@
         null,
```
Comment (dpickett) [inline coverage/.resultset.json]: you should gitignore everything in `coverage`

File: db/seeds.rb [hunk 12]
```diff
@@ -5,3 +5,21 @@
...
+  voter.address2 = voter_data[1]
+  voter.city = voter_data[2]
+  voter.state = voter_data[3]
+  voter.zip = "0" + voter_data[4]
+  voter.first_name = Faker::Name.first_name
+  voter.last_name = Faker::Name.last_name
+  elapsed = Time.now - start
+  sleep([0.21 - elapsed, 0].max)
+  voter.geocode
+  puts "#{voter.full_name} created" if voter.save
```
Comment (dpickett) [inline db/seeds.rb]: what happens if voter creation fails?

File: app/controllers/lists_controller.rb (hunk 1 above)
Comment (firechicago) [inline app/controllers/lists_controller.rb]: Definitely on the to-do list, but I think I'm going to do it as a separate pull request some time in the next day or two
